File: app/src/renderer/planio.ts
// Export and Import.
//
// The data lives on this machine already — the server listens on loopback and nothing leaves it —
// but "on this machine" is not a backup. Export writes a plain JSON file through a real Save
// dialog; Import reads one back through `import_plan`.
//
// Export is a pure read: the renderer already holds every row through its subscription, so it
// serialises what it has. Import cannot be pure, so it is one reducer and one transaction: a file
// with a bad row on line 400 leaves the existing plan exactly as it was.
//
// Ids are written as JSON numbers. They are small `auto_inc` counters, nowhere near 2^53, and
// `import_plan` remaps them on the way in anyway — the numbers in the file are only there to wire
// `laneId`, `predecessorId` and the assignment pairs back together.

import type { PlannerApp } from './types';

const FORMAT = 'vertico-planner-plan';
const VERSION = 1;

interface PlanFile {
  format: string;
  version: number;
  exportedAt: string;
  people: { id: number; name: string; role: string; avatarColour: string; initials: string }[];
  lanes: { id: number; name: string; colour: string; sortOrder: number }[];
  tasks: {
    id: number;
    laneId: number;
    name: string;
    startDay: number;
    durationDays: number;
    percentComplete: number;
    predecessorId: number;
  }[];
  assignments: { taskId: number; personId: number }[];
}

export function serialisePlan(app: PlannerApp): string {
  const snap = app.snapshot;
  const file: PlanFile = {
    format: FORMAT,
    version: VERSION,
    exportedAt: new Date().toISOString(),
    people: snap.people.map(p => ({
      id: Number(p.id),
      name: p.name,
      role: p.role,
      avatarColour: p.avatarColour,
      initials: p.initials,
    })),
    lanes: snap.lanes.map(l => ({
      id: Number(l.id),
      name: l.name,
      colour: l.colour,
      sortOrder: l.sortOrder,
    })),
    tasks: snap.tasks.map(t => ({
      id: Number(t.id),
      laneId: Number(t.laneId),
      name: t.name,
      startDay: t.startDay,
      durationDays: t.durationDays,
      percentComplete: t.percentComplete,
      predecessorId: Number(t.predecessorId),
    })),
    assignments: snap.assignments.map(a => ({
      taskId: Number(a.taskId),
      personId: Number(a.personId),
    })),
  };
  return `${JSON.stringify(file, null, 2)}\n`;
}

export async function exportPlan(app: PlannerApp): Promise<void> {
  const bridge = window.planner;
  if (!bridge) {
    app.say('The file bridge is missing — this window was not opened by the app.', 'error');
    return;
  }
  const stamp = new Date().toISOString().slice(0, 10);
  const result = await bridge.exportJson(`vertico-plan-${stamp}.json`, serialisePlan(app));
  if (result.error) app.say(`Could not write the file: ${result.error}`, 'error');
  else if (result.saved) app.say(`Exported to ${result.path}`);
}

export async function importPlan(app: PlannerApp): Promise<void> {
  const bridge = window.planner;
  if (!bridge) {
    app.say('The file bridge is missing — this window was not opened by the app.', 'error');
    return;
  }
  if (!app.conn) {
    app.say('Not connected — there is nowhere to import to.', 'error');
    return;
  }

  const result = await bridge.importJson();
  if (result.error) {
    app.say(`Could not read the file: ${result.error}`, 'error');
    return;
  }
  if (!result.opened || !result.json) return;

  let parsed: unknown;
  try {
    parsed = JSON.parse(result.json);
  } catch (error) {
    app.say(`That file is not JSON: ${(error as Error).message}`, 'error');
    return;
  }

  const file = parsed as Partial<PlanFile>;
  if (
    !file ||
    !Array.isArray(file.people) ||
    !Array.isArray(file.lanes) ||
    !Array.isArray(file.tasks) ||
    !Array.isArray(file.assignments)
  ) {
    app.say('That file is not a plan: it needs people, lanes, tasks and assignments.', 'error');
    return;
  }
  if (file.format && file.format !== FORMAT) {
    app.say(`That file says it is “${file.format}”, not a ${FORMAT}.`, 'error');
    return;
  }

  const summary = `${file.people.length} people, ${file.lanes.length} lanes, ${file.tasks.length} tasks`;
  if (!window.confirm(`Replace the whole plan with ${summary}?\n\nThis cannot be undone.`)) return;

  const ok = await app.call(() =>
    app.conn!.reducers.importPlan({
      people: file.people!.map(p => ({
        id: BigInt(p.id ?? 0),
        name: String(p.name ?? ''),
        role: String(p.role ?? ''),
        avatarColour: String(p.avatarColour ?? '#3ee8b0'),
        initials: String(p.initials ?? ''),
      })),
      lanes: file.lanes!.map(l => ({
        id: BigInt(l.id ?? 0),
        name: String(l.name ?? ''),
        colour: String(l.colour ?? '#3ee8b0'),
        sortOrder: Number(l.sortOrder ?? 0),
      })),
      tasks: file.tasks!.map(t => ({
        id: BigInt(t.id ?? 0),
        laneId: BigInt(t.laneId ?? 0),
        name: String(t.name ?? ''),
        startDay: Number(t.startDay ?? 0),
        durationDays: Number(t.durationDays ?? 1),
        percentComplete: Number(t.percentComplete ?? 0),
        predecessorId: BigInt(t.predecessorId ?? 0),
      })),
      assignments: file.assignments!.map(a => ({
        taskId: BigInt(a.taskId ?? 0),
        personId: BigInt(a.personId ?? 0),
      })),
    })
  );

  if (ok) {
    app.select({ kind: 'none' });
    app.say(`Imported ${summary}. Ids were reissued.`);
  }
}

/** The New plan button: empties every table. */
export async function wipePlan(app: PlannerApp): Promise<void> {
  if (!app.conn) {
    app.say('Not connected — nothing was changed.', 'error');
    return;
  }
  if (
    !window.confirm('Empty the whole plan — people, lanes, tasks and assignments?\n\nThis cannot be undone. Export first if you want it back.')
  ) {
    return;
  }
  const ok = await app.call(() => app.conn!.reducers.wipePlan({}));
  if (ok) {
    app.select({ kind: 'none' });
    app.say('The plan is empty.');
  }
}

File: module/src/reducers/plan_io.rs
//! Whole-plan writes: **New plan** and **Import**.
//!
//! The app's Export is a pure read — the renderer already holds every row through its
//! subscription, so it serialises what it has and hands it to a Save dialog. Import is the
//! opposite and cannot be pure, so it lives here as one reducer.
//!
//! ## Why import remaps ids instead of restoring them
//!
//! An exported file carries the ids it was written with. Re-inserting those verbatim would put
//! rows beside an `#[auto_inc]` sequence that never saw them, and the next ordinary insert would
//! eventually collide. So `import_plan` inserts every row with `id = 0`, keeps an
//! old→new map per table, and rewrites `lane_id`, `predecessor_id`, `task_id` and `person_id`
//! through it. The plan comes back identical in shape; only the numbers change.
//!
//! ## Why refusing part-way is safe
//!
//! A reducer is one transaction. Every refusal below rolls the whole import back, so a file with
//! one bad colour on row 400 leaves the existing plan exactly as it was rather than half
//! replaced.

use std::collections::BTreeMap;

use spacetimedb::{reducer, ReducerContext, SpacetimeType, Table};

use crate::reducers::require_authenticated;
use crate::tables::*;
use crate::validate;

#[derive(SpacetimeType, Clone, Debug)]
pub struct PersonImport {
    pub id: u64,
    pub name: String,
    pub role: String,
    pub avatar_colour: String,
    pub initials: String,
}

#[derive(SpacetimeType, Clone, Debug)]
pub struct LaneImport {
    pub id: u64,
    pub name: String,
    pub colour: String,
    pub sort_order: i32,
}

#[derive(SpacetimeType, Clone, Debug)]
pub struct TaskImport {
    pub id: u64,
    pub lane_id: u64,
    pub name: String,
    pub start_day: i32,
    pub duration_days: i32,
    pub percent_complete: i32,
    pub predecessor_id: u64,
}

#[derive(SpacetimeType, Clone, Debug)]
pub struct AssignmentImport {
    pub task_id: u64,
    pub person_id: u64,
}

/// Empty every table. The **New plan** command, and the first half of an import.
#[reducer]
pub fn wipe_plan(ctx: &ReducerContext) -> Result<(), String> {
    require_authenticated(ctx)?;
    wipe(ctx);
    spacetimedb::log::info!("plan wiped");
    Ok(())
}

fn wipe(ctx: &ReducerContext) {
    let ids: Vec<u64> = ctx.db.assignment().iter().map(|r| r.id).collect();
    for id in ids {
        ctx.db.assignment().id().delete(id);
    }
    let ids: Vec<u64> = ctx.db.task().iter().map(|r| r.id).collect();
    for id in ids {
        ctx.db.task().id().delete(id);
    }
    let ids: Vec<u64> = ctx.db.lane().iter().map(|r| r.id).collect();
    for id in ids {
        ctx.db.lane().id().delete(id);
    }
    let ids: Vec<u64> = ctx.db.person().iter().map(|r| r.id).collect();
    for id in ids {
        ctx.db.person().id().delete(id);
    }
}

/// Replace the whole plan with the contents of an exported file.
#[reducer]
pub fn import_plan(
    ctx: &ReducerContext,
    people: Vec<PersonImport>,
    lanes: Vec<LaneImport>,
    tasks: Vec<TaskImport>,
    assignments: Vec<AssignmentImport>,
) -> Result<(), String> {
    require_authenticated(ctx)?;
    wipe(ctx);

    let mut person_ids: BTreeMap<u64, u64> = BTreeMap::new();
    for (index, row) in people.iter().enumerate() {
        let name = validate::name(&row.name, "Person")
            .map_err(|e| format!("people[{index}]: {e}"))?;
        let role = validate::optional_text(&row.role, "Role")
            .map_err(|e| format!("people[{index}]: {e}"))?;
        let avatar_colour = validate::colour(&row.avatar_colour, "Avatar colour")
            .map_err(|e| format!("people[{index}]: {e}"))?;
        let initials = validate::initials(&row.initials, &name)
            .map_err(|e| format!("people[{index}]: {e}"))?;
        let inserted = ctx.db.person().insert(Person {
            id: 0,
            name,
            role,
            avatar_colour,
            initials,
        });
        person_ids.insert(row.id, inserted.id);
    }

    let mut lane_ids: BTreeMap<u64, u64> = BTreeMap::new();
    for (index, row) in lanes.iter().enumerate() {
        let name = validate::name(&row.name, "Lane").map_err(|e| format!("lanes[{index}]: {e}"))?;
        let colour = validate::colour(&row.colour, "Lane colour")
            .map_err(|e| format!("lanes[{index}]: {e}"))?;
        let sort_order =
            validate::sort_order(row.sort_order).map_err(|e| format!("lanes[{index}]: {e}"))?;
        let inserted = ctx.db.lane().insert(Lane {
            id: 0,
            name,
            colour,
            sort_order,
        });
        lane_ids.insert(row.id, inserted.id);
    }

    // Tasks land with no predecessor, then a second pass wires them: a file is free to list a
    // task before the one it depends on.
    let mut task_ids: BTreeMap<u64, u64> = BTreeMap::new();
    for (index, row) in tasks.iter().enumerate() {
        let name = validate::name(&row.name, "Task").map_err(|e| format!("tasks[{index}]: {e}"))?;
        let start_day =
            validate::start_day(row.start_day).map_err(|e| format!("tasks[{index}]: {e}"))?;
        let duration_days = validate::duration_days(row.duration_days)
            .map_err(|e| format!("tasks[{index}]: {e}"))?;
        let percent_complete = validate::percent(row.percent_complete)
            .map_err(|e| format!("tasks[{index}]: {e}"))?;
        validate::span(start_day, duration_days).map_err(|e| format!("tasks[{index}]: {e}"))?;
        let lane_id = *lane_ids.get(&row.lane_id).ok_or_else(|| {
            format!(
                "tasks[{index}] \"{}\" names lane {}, which the file does not contain",
                row.name, row.lane_id
            )
        })?;
        let inserted = ctx.db.task().insert(Task {
            id: 0,
            lane_id,
            name,
            start_day,
            duration_days,
            percent_complete,
            predecessor_id: 0,
        });
        task_ids.insert(row.id, inserted.id);
    }

    for (index, row) in tasks.iter().enumerate() {
        if row.predecessor_id == 0 {
            continue;
        }
        let new_id = task_ids[&row.id];
        let predecessor_id = *task_ids.get(&row.predecessor_id).ok_or_else(|| {
            format!(
                "tasks[{index}] \"{}\" names predecessor {}, which the file does not contain",
                row.name, row.predecessor_id
            )
        })?;
        if predecessor_id == new_id {
            return Err(format!(
                "tasks[{index}] \"{}\" is its own predecessor",
                row.name
            ));
        }
        let existing = ctx.db.task().id().find(new_id).expect("just inserted");
        ctx.db.task().id().update(Task {
            predecessor_id,
            ..existing
        });
    }

    // The wiring pass above can only build a cycle out of rows the file itself contains, so the
    // check runs once at the end over the finished table rather than per row.
    reject_cycles(ctx)?;

    let mut seen: Vec<(u64, u64)> = Vec::new();
    for (index, row) in assignments.iter().enumerate() {
        let task_id = *task_ids.get(&row.task_id).ok_or_else(|| {
            format!(
                "assignments[{index}] names task {}, which the file does not contain",
                row.task_id
            )
        })?;
        let person_id = *person_ids.get(&row.person_id).ok_or_else(|| {
            format!(
                "assignments[{index}] names person {}, which the file does not contain",
                row.person_id
            )
        })?;
        if seen.contains(&(task_id, person_id)) {
            continue;
        }
        seen.push((task_id, person_id));
        ctx.db.assignment().insert(Assignment {
            id: 0,
            task_id,
            person_id,
        });
    }

    spacetimedb::log::info!(
        "imported {} people, {} lanes, {} tasks, {} assignments",
        people.len(),
        lanes.len(),
        tasks.len(),
        seen.len()
    );
    Ok(())
}

/// Walk every task's predecessor chain; refuse if any of them fails to terminate.
fn reject_cycles(ctx: &ReducerContext) -> Result<(), String> {
    let guard = ctx.db.task().count() as usize + 1;
    for task in ctx.db.task().iter() {
        let mut walker = task.predecessor_id;
        let mut steps = 0usize;
        while walker != 0 {
            if walker == task.id {
                return Err(format!(
                    "The imported plan has a circular dependency around task \"{}\"",
                    task.name
                ));
            }
            let Some(next) = ctx.db.task().id().find(walker) else {
                break;
            };
            walker = next.predecessor_id;
            steps += 1;
            if steps > guard {
                return Err("The imported plan has a circular dependency".to_string());
            }
        }
    }
    Ok(())
}
